fms_extras/utils/generation.py

```diff
@@ -3,9 +3,623 @@
+    for parent_index, child_sequence_ids in enumerate(child_sequence_ids_list):
+        best_index = int(best_guess[parent_index].item())
+
+        # free all bad candidates
+        kv_cache_manager.free_sequences(
+            child_sequence_ids[:best_index] + child_sequence_ids[best_index + 1 :]
+        )
+
+        # decrease the context length of the sequence which used to be sequence length + inp_len by the number of incorrect tokens
+        # for the best candidate
+        best_sequence_id = child_sequence_ids[best_index]
+        parent_sequence_ids.append(best_sequence_id)
+        kv_cache_manager.remove_tokens(
+            best_sequence_id, decode_seq_length - n_correct[parent_index].item() - 1  # type: ignore
+        )
+    return parent_sequence_ids
+
+
+def __apply_batch_expansion_to_cache_metadata(
+    cache_data: PagedAttentionCacheData,
+):
+    """
+    When performing paged attention, the kernels require that the sequence length be of size 1. In typical generation,
+    this is fine as only 1 token is ever being sent into the model for each sequence in the batch, however for
+    speculative decoding, n tokens are sent for each sequence in the batch where n is the number of predictions.
+
+    To properly run paged attention with the larger sequence length, we must expand the sequence into the batch
+    dimension, and modify the cache metadata to reflect this.
+
+    example:
+
+    model_input_lengths: [4, 4]
+
+    before expansion:
+
+    block_mapping -
+
+    [
+      [1, 2],
+      [10, 11, 12, 13]
+    ]
+
+    context_lengths -
+
+    [22, 66]
+
+    after expansion:
+
+    block_mapping -
+
+    [
+        [1, 2],
+        [1, 2],
+        [1, 2],
+        [1, 2],
+        [10, 11, 12],
+        [10, 11, 12],
+        [10, 11, 12, 13],
+        [10, 11, 12, 13],
+    ]
+
+    context_lengths -
+
+    [19, 20, 21, 22, 63, 64, 65, 66]
+
+    Args:
+        cache_data: PagedAttentionCacheData
+            the cache data created after allocation
+    """
+    # Set up kv cache metadata for paged memory access over tokens/candidates with shared prefixes
+    context_lengths = cache_data.context_lengths  # bk
+    inflate_factor = (
+        cache_data.query_length
+        if cache_data.unflatten_indices is None
+        else cache_data.unflatten_indices.size(-1)
+    )
+    # no reason to check type here as generation allocation always returns context_lengths
+    context_lengths = context_lengths.unsqueeze(1).expand(  # type: ignore
+        -1, inflate_factor
+    )  # bk n
+    # subtract arange(inp_len) to get lengths for each token in candidates
+    cache_data.context_lengths = (
+        context_lengths.sub(context_lengths.sign().cumsum(1).flip([1]).sub(1))
+        .int()
+        .view(-1)
+    )  # bkn
+    cache_data.block_mapping = cache_data.block_mapping.repeat_interleave(
+        inflate_factor, dim=0
+    )  # bkn n_blocks
+
+
+def __apply_flattening_to_cache_metadata(
```

Comment:
since this is operating on the internal state of cache_data, maybe it should be a member function on PagedAttentionCacheData instead of helper function here?

Reply:
fixed in latest commit